lib/charms/opensearch/v0/opensearch_health.py

```diff
@@ -118,37 +119,38 @@ def apply_for_unit(self, status: str, host: Optional[str] = None):
         response: Optional[Dict[str, any]] = None
-        if wait_for_green_first:
-            try:
-                response = ClusterState.health(
-                    self._opensearch,
-                    True,
-                    host=host,
-                    alt_hosts=self._charm.alt_hosts,
-                )
-            except OpenSearchHttpError:
-                # it timed out, settle with current status, fetched next without the 1min wait
-                pass
-
-        if not response:
-            response = ClusterState.health(
-                self._opensearch,
-                False,
-                host=host,
-                alt_hosts=self._charm.alt_hosts,
-            )
-
-        if not response:
-            return None
-
-        logger.info(f"Health: {response['status']}")
-        status = response["status"].lower()
-        if status != HealthColors.YELLOW:
-            return status
 
-        # we differentiate between a temp yellow (moving shards) and perm one (missing replicas)
-        shards_by_state = ClusterState.shards_by_state(
-            self._opensearch, host=host, alt_hosts=self._charm.alt_hosts
-        )
-        busy_shards = shards_by_state.get("INITIALIZING", 0) + shards_by_state.get("RELOCATING", 0)
-        return HealthColors.YELLOW_TEMP if busy_shards > 0 else HealthColors.YELLOW
+        try:
+            for attempt in Retrying(stop=stop_after_attempt(5), wait=wait_fixed(5)):
+                with attempt:
+                    response = ClusterState.health(
+                        self._opensearch,
+                        False,
+                        host=host,
+                        alt_hosts=self._charm.alt_hosts,
+                    )
+
+                    logger.info(f"Health: {response}")
+                    status = response["status"].lower()
+                    # If we must wait for green
+                    if wait_for_green_first and status == HealthColors.GREEN:
+                        return status
+                    elif wait_for_green_first:
+                        raise Exception()  # retry
+
+                    if status in [HealthColors.RED, HealthColors.UNKNOWN]:
+                        # Status may be int or different strs, make sure they fit RED or UNKNOWN
+                        return status
+
+                    # we differentiate between a temp yellow (moving shards)
+                    # and perm one (missing replicas)
+                    shards_by_state = ClusterState.shards_by_state(
+                        self._opensearch, host=host, alt_hosts=self._charm.alt_hosts
+                    )
+                    busy_shards = shards_by_state.get("INITIALIZING", 0) + shards_by_state.get(
+                        "RELOCATING", 0
+                    )
+                    return HealthColors.YELLOW_TEMP if busy_shards > 0 else HealthColors.YELLOW
+
+        except RetryError:
+            return HealthColors.UNKNOWN
```

Comment:
Proposed fix for: https://github.com/canonical/opensearch-operator/pull/130#issuecomment-1744451875